src/libsyntax/parse/lexer/mod.rs

```diff
@@ -1367,64 +1319,124 @@ impl<'a> StringReader<'a> {
+                let last_bpos = self.pos;
+                self.fatal_span_char(start_bpos,
+                                     last_bpos,
+                                     "found invalid character; only `#` is allowed \
+                                      in raw string delimitation",
+                                     c).raise();
+            }
         }
+
         self.bump();
         let content_start_bpos = self.pos;
         let mut content_end_bpos;
+        let mut valid = true;
+        let mut spans = vec![];
+
         'outer: loop {
-            match self.ch {
-                None => {
-                    self.fail_unterminated_raw_string(start_bpos, hash_count);
-                }
-                Some('"') => {
+            match (self.ch, raw_type) {
+                (None, _) => {
+                    spans.push(self.mk_sp(self.pos, self.pos));
+                    self.fail_unterminated_raw_string(bpos_span, hash_count, spans);
+                },
+                (Some('"'), _) => {
                     content_end_bpos = self.pos;
                     for _ in 0..hash_count {
                         self.bump();
                         if !self.ch_is('#') {
+                            spans.push(self.mk_sp(content_end_bpos, self.pos));
                             continue 'outer;
                         }
                     }
                     break;
                 }
-                Some(c) => {
+                (Some('\r'), RawStringType::Unicode) => {
+                    if !self.nextch_is('\n') {
+                        let last_bpos = self.pos;
+                        self.err_span_(start_bpos,
+                                       last_bpos,
+                                       "bare CR not allowed in raw string, use \\r \
+                                        instead");
+                        valid = false;
+                    }
+                }
+                (Some(c), RawStringType::Byte) => {
                     if c > '\x7F' {
                         let pos = self.pos;
                         self.err_span_char(pos, pos, "raw byte string must be ASCII", c);
                     }
                 }
+                _ => (),
             }
             self.bump();
         }
 
         self.bump();
+        if self.ch_is('#') {
+            let lo = self.pos;
+            while self.ch_is('#') {
+                self.bump();
+            }
+
+            let sp = self.mk_sp(start_bpos, self.pos);
+            let sp_beg = self.mk_sp(BytePos(start_bpos.0 + 1),
+                                    BytePos(start_bpos.0 + 1 + hash_count as u32));
+            let sp_end = self.mk_sp(BytePos(lo.0 - hash_count as u32), self.pos);
+
+            let mut err = self.sess
+                .span_diagnostic.struct_span_err(sp, "too many `#` when terminating raw string");
+            err.span_label(sp_beg, format!("The raw string has {} leading `#`...", hash_count));
```

Comment:
Do not capitalize sentences unless it's a paragraph

Say "this raw string has..."